trajectory_tracker/test/src/test_path2d.cpp

```diff
@@ -144,6 +148,163 @@ TEST(Path2D, LocalGoalWithSwitchBack)
+    EXPECT_EQ(nearest_with_dist.first, path.begin() + 2);
+    EXPECT_NEAR(nearest_with_dist.second, std::sqrt(std::pow(1.0 - 0.7, 2) + std::pow(0.5 - 0.5, 2)), 1.0e-6);
+  }
+  {
+    // Test edge cases
+    const auto nearest_with_dist =
+        path.findNearestWithDistance(path.begin() + 5, path.begin() + 5, Eigen::Vector2d(1.0, 0.5));
+    EXPECT_EQ(nearest_with_dist.first, path.begin() + 5);
+    EXPECT_NEAR(nearest_with_dist.second, std::sqrt(std::pow(1.0 - 0.9, 2) + std::pow(0.5 - 0.8, 2)), 1.0e-6);
+
+    const auto invalid_result =
+        path.findNearestWithDistance(path.end(), path.end(), Eigen::Vector2d(1.0, 0.5));
+    EXPECT_EQ(invalid_result.first, path.end());
+    EXPECT_EQ(invalid_result.second, std::numeric_limits<double>::max());
+  }
+}
+
+TEST(Path2D, Conversions)
+{
+  nav_msgs::Path path_msg_org;
+  path_msg_org.poses.resize(8);
+  path_msg_org.poses[0].pose.position.x = 0.0;
+  path_msg_org.poses[0].pose.position.y = 0.0;
+  path_msg_org.poses[0].pose.orientation.w = 1.0;
+  path_msg_org.poses[1].pose.position.x = 1.0;  // Start of in-place turning
+  path_msg_org.poses[1].pose.position.y = 0.0;
+  path_msg_org.poses[1].pose.orientation.w = 1.0;
+  path_msg_org.poses[2].pose.position.x = 1.0;
+  path_msg_org.poses[2].pose.position.y = 0.0;
+  path_msg_org.poses[2].pose.orientation.z = std::sin(M_PI / 8);
+  path_msg_org.poses[2].pose.orientation.w = std::cos(M_PI / 8);
+  path_msg_org.poses[3].pose.position.x = 1.0;  // End of in-place turning
+  path_msg_org.poses[3].pose.position.y = 0.0;
+  path_msg_org.poses[3].pose.orientation.z = std::sin(M_PI / 4);
+  path_msg_org.poses[3].pose.orientation.w = std::cos(M_PI / 4);
+  path_msg_org.poses[4].pose.position.x = 1.0;
+  path_msg_org.poses[4].pose.position.y = 1.0;
+  path_msg_org.poses[4].pose.orientation.z = std::sin(M_PI / 4);
+  path_msg_org.poses[4].pose.orientation.w = std::cos(M_PI / 4);
+  path_msg_org.poses[5].pose.position.x = 1.0;  // Start of in-place turning
+  path_msg_org.poses[5].pose.position.y = 2.0;
+  path_msg_org.poses[5].pose.orientation.z = std::sin(M_PI / 4);
+  path_msg_org.poses[5].pose.orientation.w = std::cos(M_PI / 4);
+  path_msg_org.poses[6].pose.position.x = 1.0;
+  path_msg_org.poses[6].pose.position.y = 2.0;
+  path_msg_org.poses[6].pose.orientation.z = std::sin(M_PI / 4 + M_PI / 6);
+  path_msg_org.poses[6].pose.orientation.w = std::cos(M_PI / 4 + M_PI / 6);
+  path_msg_org.poses[7].pose.position.x = 1.0;  // End of in-place turning
+  path_msg_org.poses[7].pose.position.y = 2.0;
+  path_msg_org.poses[7].pose.orientation.z = std::sin(M_PI / 4 + M_PI / 3);
+  path_msg_org.poses[7].pose.orientation.w = std::cos(M_PI / 4 + M_PI / 3);
+
+  trajectory_tracker::Path2D path;
+  path.fromMsg(path_msg_org);
+
+  ASSERT_EQ(path.size(), 6);
+  for (size_t i = 0; i < path.size(); ++i)
+  {
+    size_t org_idx = i;
+    if (i == 5)
+    {
+      org_idx += 2;
+    }
+    else if (i >= 2)
+    {
+      org_idx += 1;
+    }
```

Comment:
Defining a array to have a list of corresponding `org_idx` would be more readable